Task-36 Экраны выбора места работы

IvanPodlesnykh opened the pull request:
1. Реализовал главный экран выбора места работы
2. Реализовал экран выбора страны
3. Реализовал экран выбора региона
4. Исправил HHApi, интерактор и репозиторий для работы с поиском по Area
5. Исправил Area, Mapper, AreaResponse, AreaRequest

IvanPodlesnykh committed:
Task-36 Экраны выбора места работы
1. Реализовал главный экран выбора места работы
2. Реализовал экран выбора страны
3. Реализовал экран выбора региона
4. Исправил HHApi, интерактор и репозиторий для работы с поиском по Area
5. Исправил Area, Mapper, AreaResponse, AreaRequest

IvanPodlesnykh committed:
Task-36 Экраны выбора места работы
Правки для detekt

bukholovdmitry commented on app/src/main/java/ru/practicum/android/diploma/presentation/workplace/WorkplaceViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+        var newArea: Area = region
+
+        while (!newArea.parentId.isNullOrEmpty()) {
+            dictionariesInteractor.getAreasById(newArea.parentId!!).collect {
```
В плейлист мейкере ревьюр сделал замечание что следует избегать опасных вызовов с !!

IvanPodlesnykh replied:
> В плейлист мейкере ревьюр сделал замечание что следует избегать опасных вызовов с !!

Тут в while идет проверка на нуль этого параметра, так что не должно быть проблем, но я подумаю как можно переделать

herrfreeman replied:
Есть проверка на null, по этому !! допустимо

bukholovdmitry commented on app/src/main/java/ru/practicum/android/diploma/ui/country/CountryFragment.kt:
```diff
@@ -23,10 +35,64 @@ class CountryFragment : Fragment() {
+        binding.errorText.isVisible = false
+
+        val adapter = AreaAdapter {
+            parent.workplaceViewModel.setCountry(it)
```
Такое обращение к фрагменту родителю наверно не есть ок, лучше использовать интерактор для сохранения выбранного значения и при возвращении во фрагмент проверять есть ли новое значение в sharedPreferences

IvanPodlesnykh replied:
> Такое обращение к фрагменту родителю наверно не есть ок, лучше использовать интерактор для сохранения выбранного значения и при возвращении во фрагмент проверять есть ли новое значение в sharedPreferences

Тут я проверил, приложение падает, начал исправлять и избавляться от parentFragment. Фрагмент CountryFragment уже исправил

IvanPodlesnykh committed:
Task-36 Экраны выбора места работы
Добавил недостающий LayoutManager для списков стран и регионов

IvanPodlesnykh committed:
Task-36 Экраны выбора места работы
Исправил ошибку сохранения страны

lodrean commented on app/src/main/res/layout/fragment_workplace.xml:
```diff
@@ -23,4 +23,122 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
+
```
надо посмотреть как это в работе, у меня пока не получилось фрагмент запустить, возможно замерджить надо и проверить. я наверное потом их рефакторинг сделаю, у меня все через TextInputLayout реализовано

bukholovdmitry reviewed (approved):
Переход в страны работает, а работы с переходом в регионы надо тестить вместе с логикой сохранения.

IvanPodlesnykh committed:
Task-36 Экраны выбора места работы
Исправлен выбор региона
Исправлены классы по получению из сети конкретного региона (по id)
Исправлена работа основного экрана фильтра мест работы

TODO: Поправить работу фильтра

herrfreeman commented on app/src/main/java/ru/practicum/android/diploma/presentation/workplace/WorkplaceViewModel.kt:
```diff
@@ -0,0 +1,68 @@
+
+        while (!newArea.parentId.isNullOrEmpty()) {
+            dictionariesInteractor.getAreasById(newArea.parentId!!).collect {
+                newArea = (it as Resource.Success).data!!
```
А вот тут мы получается не проверяем ошибки, надо будет потом исправить

herrfreeman commented on app/src/main/java/ru/practicum/android/diploma/ui/adapters/AreaAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+import ru.practicum.android.diploma.ui.viewholders.AreaViewHolder
+import ru.practicum.android.diploma.util.OnAreaItemClickListener
+
+class AreaAdapter(private val onAreaItemClickListener: OnAreaItemClickListener) : Adapter<AreaViewHolder>() {
```
Adapter эквивалентно RecyclerView.Adapter ?

IvanPodlesnykh replied:
> Adapter эквивалентно RecyclerView.Adapter ?

Да, просто импортируется RecyclerView.Adapter, а не RecyclerView

herrfreeman commented on app/src/main/java/ru/practicum/android/diploma/ui/country/CountryFragment.kt:
```diff
@@ -23,10 +32,64 @@ class CountryFragment : Fragment() {
+
+        val adapter = AreaAdapter {
+            viewModel.setCountry(it)
+            findNavController().navigateUp()
```
Почему Up а не назад?
Мне кажется адаптер нужно вынести в свойства фрагмента

IvanPodlesnykh replied:
> Почему Up а не назад? Мне кажется адаптер нужно вынести в свойства фрагмента

Читал, что при навигации в рамках одного приложения предпочтительнее использовать navigateUp, а не popBackStack

herrfreeman commented on app/src/main/java/ru/practicum/android/diploma/ui/country/CountryFragment.kt:
```diff
@@ -23,10 +32,64 @@ class CountryFragment : Fragment() {
+
+    override fun onResume() {
+        super.onResume()
+        viewModel.loadCountriesList()
```
Есть смысл каждый раз при активации перезагружать страны? Может перенести в инит вью модели?

herrfreeman commented on app/src/main/java/ru/practicum/android/diploma/presentation/country/CountryViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+
+    fun setCountry(country: Area) {
+        filterInteractor.setArea(country)
+        filterInteractor.apply()
```
```suggestion
```
Нужно убрать, так как это применяет фильтр

herrfreeman commented on app/src/main/java/ru/practicum/android/diploma/presentation/region/RegionViewModel.kt:
```diff
@@ -0,0 +1,153 @@
+    private val dictionariesInteractor: DictionariesInteractor
+) : ViewModel() {
+
+    private val regionLiveData = MutableLiveData<RegionState>()
```
Желательно чтобы LiveData была одна на вью модель

IvanPodlesnykh replied:
> Желательно чтобы LiveData была одна на вью модель

Тут одна LiveData для списка регионов и реакции на его изменения (например при поиске), а вторая LiveData для получения информации о том, какой сейчас фильтр применен: например если уже выбрана страна, то в списке регионов не появится список регионов из других стран.

herrfreeman reviewed (approved):
Критических замечаний нет. На экране региона хорошо бы сделать единые стейты

herrfreeman committed:
Merge branch 'develop' into task-36
